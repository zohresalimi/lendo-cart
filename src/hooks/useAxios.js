import { useState, useEffect } from "react";
import axios from "axios";

export default function useAxios() {
  const [data, setData] = useState([]);
  const [url, setUrl] = useState("api/products");
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const fetchData = async () => {
      setIsError(false);
      setIsLoading(true);

      try {
        const result = await axios(url);

        setData(result.data);
      } catch (error) {
        setIsError(true);
      }

      setIsLoading(false);
    };

    fetchData();
  }, [url]);

  return [{ data, isLoading, isError }, setUrl];
}
