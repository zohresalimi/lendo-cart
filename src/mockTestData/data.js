import React, { Children, useReducer } from "react";
import { BrowserRouter as Router } from "react-router-dom";

import AppContext from "../store/context";
import reducer from "../store/reducer";

export const getTestStore = () => {
  return {
    currentProduct: {
      id: 1,
      selectedFeatures: {
        color: "whith",
        power: 6.5,
      },
    },
    products: {
      byColor: {
        white: {
          1: { power: [6.5, 9.5], remaining: 3 },
          2: { power: [6.5, 9.5], remaining: 3 },
        },
        red: {
          1: { power: [6.5, 9.5], remaining: 7 },
          2: { power: [6.5, 9.5], remaining: 7 },
        },
      },
      byId: {
        1: {
          id: 1,
          name: "Philips hue bulb",
          brand: "Philips",
          price: "500",
          available: true,
          weight: 0.2,
          options: [
            {
              color: "white",
              power: [6.5, 9.5],
              quantity: 3,
            },
            {
              color: "red",
              power: [6.5, 9.5],
              quantity: 7,
            },
          ],
        },
        2: {
          id: 2,
          name: "TrÃ¥dfria Lampor",
          brand: "IKEA",
          price: "300",
          available: true,
          weight: 0.2,
          options: [
            {
              color: "white",
              power: [6.5, 9.5],
              quantity: 3,
            },
            {
              color: "red",
              power: [6.5, 9.5],
              quantity: 7,
            },
            {
              color: "green",
              power: [6.5],
              quantity: 0,
            },
          ],
        },
      },
    },
    shoppingCart: [
      { id: 1, quantity: 1, selectedFeatures: { color: "white", power: 6.5 } },
    ],
  };
};

export const WithProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, getTestStore());

  return (
    <AppContext.Provider
      value={{
        state,
        dispatch: props.mockDispatch || dispatch,
      }}
    >
      <Router>{props.children}</Router>
    </AppContext.Provider>
  );
};
