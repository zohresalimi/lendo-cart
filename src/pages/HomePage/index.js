import React, { useEffect, useContext, useState } from "react";
import { Grid } from "@material-ui/core";
import LinearProgress from "@material-ui/core/LinearProgress";

import useAxios from "../../hooks/useAxios";
import AppContext from "../../store/context";
import { SET_PRODUCTS_REDUCER } from "../../constant";

import DataTable from "../../components/DataTable";
import SnakBar from "../../components/SnakBar";
import useStyles from "./style";
import { useHistory, useLocation } from "react-router-dom";

export default function HomePage() {
  const { dispatch } = useContext(AppContext);
  const [showSnakbar, setShowSnakbar] = useState(false);
  const location = useLocation();
  const history = useHistory();

  const [{ data, isLoading, isError }] = useAxios("api/products");

  const classes = useStyles();

  useEffect(() => {
    if (data) {
      dispatch({ type: SET_PRODUCTS_REDUCER, data: data.items });
    }
  }, [data, dispatch]);

  useEffect(() => {
    if (location.state?.messageText) {
      setShowSnakbar(true);
    }
  }, [history, location.state]);

  return (
    <>
      <Grid container spacing={3}>
        <Grid item xs={12}>
          {isError && <div>Somthing went wrong</div>}
          {isLoading ? (
            <LinearProgress className={classes.progressBar} />
          ) : (
            <DataTable />
          )}
        </Grid>
      </Grid>
      <SnakBar
        visible={showSnakbar}
        hideSnakbar={() => setShowSnakbar(false)}
        text={location.state?.messageText}
      />
    </>
  );
}
