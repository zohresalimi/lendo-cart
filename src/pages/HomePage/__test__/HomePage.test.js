import * as React from "react";

import { act, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { WithProvider } from "../../../mockTestData/data";
import * as hooks from "../../../hooks/useAxios";

import HomePage from "..";

jest.mock("../../../components/DataTable", () => () => (
  <div>DataTable Component</div>
));

async function renderWrapper(props) {
  let component;

  act(() => {
    component = render(
      <WithProvider {...props}>
        <HomePage path="/" />
      </WithProvider>
    );
  });

  return component;
}

describe("Home Page", () => {
  test("shows loading if products are not loaded yet", async () => {
    const mockAxios = jest.fn();
    mockAxios.mockReturnValue([
      {
        isLoading: true,
        data: null,
        isError: false,
      },
    ]);

    jest.spyOn(hooks, "default").mockImplementation(mockAxios);

    const mockDispatch = jest.fn();

    const { container } = await renderWrapper({ mockDispatch });
    expect(mockDispatch).toHaveBeenCalledTimes(0);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("shows the product from api call", async () => {
    const mockAxios = jest.fn();
    mockAxios.mockReturnValue([
      {
        isLoading: false,
        data: {
          items: [
            {
              id: 1,
              name: "Philips hue bulb",
              brand: "Philips",
              price: "500",
            },
            {
              id: 2,
              name: "Tr√•dfria Lampor",
              brand: "IKEA",
              price: "300",
            },
          ],
        },
        isError: false,
      },
    ]);

    jest.spyOn(hooks, "default").mockImplementation(mockAxios);

    const mockDispatch = jest.fn();

    const { container } = await renderWrapper({ mockDispatch });
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(container.firstChild).toMatchSnapshot();
  });
});
