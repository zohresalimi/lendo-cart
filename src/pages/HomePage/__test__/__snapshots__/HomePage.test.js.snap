// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page shows loading if products are not loaded yet 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <div
      class="MuiLinearProgress-root MuiLinearProgress-colorPrimary makeStyles-progressBar-1 MuiLinearProgress-indeterminate"
      role="progressbar"
    >
      <div
        class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
      />
      <div
        class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
      />
    </div>
  </div>
</div>
`;

exports[`Home Page shows the product from api call 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <div>
      DataTable Component
    </div>
  </div>
</div>
`;
